scipy/optimize/_hungarian.py

```diff
@@ -50,233 +96,232 @@ def linear_sum_assignment(cost_matrix):
+        self.row_marked *= False
+        self.col_marked *= False
+        self.row_saturated *= False
+        self.col_saturated *= False
+        self.marked *= False
+        self.col_saturated_size = 0
+
+    def _min_vertex_cover(self):
+        """Find a minimum vertex cover of 0-induced graph"""
+
+        # Start with all unsaturated row vertices
+        self.row_marked = self.row_saturated == False
+
+        # We keep trying to find new vertices reachable by augmenting paths.
+        while True:
+            found = self.col_marked.sum()
+            # Saturated column neighbors of rows from previous round
+            self.col_marked = np.any(self.matrix[self.row_marked] == 0, axis=0)
+            if found == self.col_marked.sum():
+                self.col_saturated_size = found
+                return
             else:
-                col = star_col
-                state.row_uncovered[row] = False
-                state.col_uncovered[col] = True
-                covered_C[:, col] = C[:, col] * (
-                    np.asarray(state.row_uncovered, dtype=int))
-                covered_C[row] = 0
-
-
-def _step5(state):
-    """
-    Construct a series of alternating primed and starred zeros as follows.
-    Let Z0 represent the uncovered primed zero found in Step 4.
-    Let Z1 denote the starred zero in the column of Z0 (if any).
-    Let Z2 denote the primed zero in the row of Z1 (there will always be one).
-    Continue until the series terminates at a primed zero that has no starred
-    zero in its column. Unstar each starred zero of the series, star each
-    primed zero of the series, erase all primes and uncover every line in the
-    matrix. Return to Step 3
-    """
-    count = 0
-    path = state.path
-    path[count, 0] = state.Z0_r
-    path[count, 1] = state.Z0_c
-
-    while True:
-        # Find the first starred element in the col defined by
-        # the path.
-        row = np.argmax(state.marked[:, path[count, 1]] == 1)
-        if state.marked[row, path[count, 1]] != 1:
-            # Could not find one
-            break
+                # Mark rows that are matched with columns found above
+                self.row_marked[np.any(self.marked[:, self.col_marked],
+                                       axis=1)] = True
+
+    def _aug_paths(self):
+        """Find an augmenting path if one exists from maximal matching."""
+        # Check unsaturated row vertices for augmenting paths
+        for row in np.nonzero(self.row_saturated == False)[0]:
+            path_row, path_col = self._aug_path(row)
+            if not path_col:
+                continue
+            if not len(path_row + path_col) % 2:
+                for i in range(len(path_row) - 1):
+                    self.marked[path_row[i], path_col[i]] = 1
+                    self.marked[path_row[i], path_col[i + 1]] = 0
+                self.marked[path_row[-1], path_col[-1]] = 1
+                self.row_saturated[path_row[-1]] = \
```

Comment:
this looks a bit odd, can you split into two separate statements for clarity?

Reply:
Does the latest commit look more undestandable?